distributed/utils.py

```diff
@@ -1440,3 +1441,91 @@ def __getattr__(name):
         return import_term(use_instead)
     else:
         raise AttributeError(f"module {__name__} has no attribute {name}")
+
+
+class SupportsToDict(Protocol):
+    def to_dict(
+        self, *, exclude: Container[str] | None = None, **kwargs
+    ) -> dict[str, AnyType]:
+        ...
+
+
+@overload
+def recursive_to_dict(
+    obj: SupportsToDict, exclude: Container[str] = None, seen: set[AnyType] = None
+) -> dict[str, AnyType]:
+    ...
+
+
+@overload
+def recursive_to_dict(
+    obj: Sequence, exclude: Container[str] = None, seen: set[AnyType] = None
+) -> Sequence:
+    ...
+
+
+@overload
+def recursive_to_dict(
+    obj: dict, exclude: Container[str] = None, seen: set[AnyType] = None
+) -> dict:
+    ...
+
+
+@overload
+def recursive_to_dict(
+    obj: None, exclude: Container[str] = None, seen: set[AnyType] = None
+) -> None:
+    ...
```

Comment:
I trust that this is necessary to make `mypy` happy, but it's unfortunate that there's not a more compact version

Reply:
One can simplify this with many `AnyType`s but then the entire value is lost. I also want to point out that this is not the actually correct annotations since there are paths not covered by these annotations but this is sufficient for mypy

Reply:
That should rather be taken as a strong encouragement to never write code like that :)

In this particular case, the behaviour I was looking for is sufficiently weird to justify this, kind of.